### DISCUSSION TITLE
"Proper" way to call API routes?

### DISCUSSION THREAD

**Discussion by JamesSingleton** [vercel/next.js#21306] [Help]
Hey Everyone 👋🏼 ,

So I have started to look into API routes to serve some dummy data for website templates. All the examples with using `getServerSideProps` shows calling an API route like so:

```js
export async function getServerSideProps(context) {
  const res = await fetch('/api/team')
  const data = await res.json()

  return {
    props: {
      data
    }, 
  }
}
```
With your API route looking like
```js
export default (req, res) => {
  res.statusCode = 200
  res.setHeader('Content-Type', 'application/json')
  res.json([
    {
      name: 'Iron Man',
      title: 'Real Estate Agent'
    },
    {
      name: ' Deadpool',
      title: 'Real Estate Agent'
    },
    {
      name: 'Dr. Strange',
      title: 'Mortgage Consultant'
    }
  ])
}
```

However, calling it like that results in an error saying that absolute URLs are only allowed. If that is the case, why is the documentation not using absolute URLs?

What I have been told is to create an API route that looks like
```js
import getData from "../../utils/getData"

export default async function (req, res) {
  const data = await getData("/api/data")

  res.json(data)
}
```
Where `getData` looks like this (I am assuming this is where the dummy data would go)
```js
export default async function getData(origin) {
  // Note that this timeout promise is only to demonstrate/simulate network latency.
  // You could just as easily return a plain object directly!
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        origin,
        time: Date.now(),
      })
    }, 500)
  })
}
```

And then your gSSP would look like
```js
export async function getServerSideProps() {
  const initialData = await getData("getServerSideProps")

  return { props: { initialData } }
}
```

Why does that seem so _unintuitive_? Is that even the _recommended_ way?

**Comment by gomflo** [comment]
I think using `getServerSideProps` and doing an API call is not the proper way as `getServerSideProps`  runs on the server, so you can run a query directly in the `getServerSideProps`

![image](https://user-images.githubusercontent.com/141670/104985293-3c089180-59d6-11eb-8e3b-90289875bb01.png)

**Comment by jpedroschmitz** [comment]
I wouldn't consider `absolute URLs are only allowed` an error. How else would fetch know where to go on the server?

There are two options to solve this problem (that I'm aware of):

**1.** Use an environment variable and define the API URL.
```js
export async function getServerSideProps(context) {
  const res = await fetch(`${process.env.API_URL}/api/team`)
  const data = await res.json()

  return {
    props: {
      data
    }, 
  }
}
```
**2.** Use a function to get the data or use the API route code directly in `getServerSideProps/getStaticProps`.

If you won't use the API route anywhere else, I would simply get the API route code and use it on the `getServerSideProps` function. This way you avoid a `fetch` to hit the API endpoint.

If you need the API route in other places for your application, you can do [something like this](https://github.com/vercel/next.js/discussions/16068#discussioncomment-127246).
